### PR TITLE
PUP-6792,6793,6795-6797 Solaris Updates

### PR DESCRIPTION
**Pull request by shawnferry**
(PUP-6792) SMF should recursively enable services

SMF supports enabling a service and all services which that service
requires. With this patch recursive starts are the default behavior.

(PUP-6793) pkg provider should be default for Solaris 12

Puppet needs to use the pkg provider to manage packages on Solaris 12

(PUP-6795) Solaris does not support gzipped man pages

Solaris does not support gzipped man pages. Skip gzipping when
installing on Solaris.

(PUP-6796) Support setting/checking shell in user_role_add

Puppet should allow a users shell to be set in a Solaris role

(PUP-6797) service should check for smf service first

(PUP-6860) puppet on solaris 11 should support exists? for services

Attempts to manage non-existent services resulted in uninformative
errors. With this patch non-existent services return better error
messages.

### CODE REVIEW

**Comment by MosesMendoza** [comment]
@shawnferry thanks yet again for the contribution on Solaris! Any chance you'd be willing to split these into atomic commits? This is for a couple reasons - one so that reviewers can more easily review your changes and associate test changes w/code changes and second in case we end up needing to back out some piece of functionality we can cleanly revert that piece instead of having to revert all of the changes.

**Comment by shawnferry** [comment]
@MosesMendoza done, sorry for the delay I missed the comment and the label change

**Comment by shawnferry** [comment]
@MosesMendoza  any idea why this is failing? I can't find the reported string anywhere in git log output.

```
Checking commits...
(PUP-6797) service should check for smf service first
...passed
(PUP-6796) Support setting/checking shell in user_role_add
...passed
(PUP-6792) SMF should recursively enable services
...passed
(PUP-6795) Solaris does not support gzipped man pages
...passed
(PUP-6793) pkg provider should be default for Solaris 12
...passed
rake aborted!
	This commit summary didn't match CONTRIBUTING.md guidelines:
		Delete file commited to wrong repo
```

**Comment by MosesMendoza** [comment]
Sorry about that failure @shawnferry - all of your commits are correctly formatted. Our commit format check has trouble actually checking _only_ the commits in your PR and gets caught up on prior commits unrelated to your PR. I dislike it and filed PUP-6849 to see about getting it removed until we can fix it.

Thanks again for the contribution! I'll try to get this reviewed...

File: lib/puppet/provider/package/pkg.rb [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
   confine :osfamily => :solaris
 
-  defaultfor :osfamily => :solaris, :kernelrelease => '5.11'
+  defaultfor :osfamily => :solaris, :kernelrelease => ['5.11', '5.12']
```
Comment (MosesMendoza) [inline lib/puppet/provider/package/pkg.rb]: I think it would be good to get spec coverage for this change, ie it's the default for solaris 11 and 12, and not the default for solaris 10. It doesn't look like we have something checking this for 11 yet, but a similar example is here: https://github.com/puppetlabs/puppet/blob/master/spec/unit/provider/package/dnf_spec.rb#L8-L26

File: install.rb [hunk 2]
```diff
@@ -108,9 +108,12 @@ def do_man(man, strip = 'man/')
     FileUtils.makedirs(om, {:mode => 0755, :verbose => true})
     FileUtils.chmod(0755, om)
     FileUtils.install(mf, omf, {:mode => 0644, :preserve => true, :verbose => true})
-    gzip = %x{which gzip}
-    gzip.chomp!
-    %x{#{gzip} -f #{omf}}
+    # Solaris does not support gzipped man pages
+    unless Facter["os"].value["family"] == "Solaris"
```
Comment (MosesMendoza) [inline install.rb]: I think `install.rb` can actually be run without requiring that `facter` be installed by passing `--no-check-prereqs`. If that's done and facter isn't installed I think this breaks.

On line 242 we already do a check for OS after confirming we aren't in 'no-prereqs' mode and assign the result to `$operatingsystem`, and then later use it to decide if we should call `do_man`. Maybe `do_man` could be extended to include an optional param of whether to compress man pages (defaulting to true), and then when calling do_man on 480 trigger that option in the negative if we're on solaris?

File: lib/puppet/provider/service/smf.rb [hunk 3]
```diff
@@ -80,7 +80,7 @@ def startcmd
     when :maintenance, :degraded
       [command(:adm), :clear, @resource[:name]]
     else
-      [command(:adm), :enable, "-s", @resource[:name]]
+      [command(:adm), :enable, "-rs", @resource[:name]]
```
Comment (MosesMendoza) [inline lib/puppet/provider/service/smf.rb]: In some cases it seems like recursively enabling services is just the default - ie with systemd, if a service expresses dependencies those dependencies are also run on start. Do you know if there's any specific reason this isn't just the default with solaris? ie are you aware of any reason why this change would cause unexpected issues/breakage for anyone in solaris land?

File: lib/puppet/provider/package/pkg.rb (hunk 1 above)
Comment (shawnferry) [inline lib/puppet/provider/package/pkg.rb]: done

File: lib/puppet/provider/service/smf.rb (hunk 3 above)
Comment (shawnferry) [inline lib/puppet/provider/service/smf.rb]: We committed the original behavior and it's unlikely that we will ever change it. We leave it up to the user to choose the behavior they want when manually executing. However, we think it should be the default behavior for puppet.

File: install.rb (hunk 2 above)
Comment (shawnferry) [inline install.rb]: Would you think to add a --[no]compress-man or similar option?

I can very easily modify my invocation of install.rb to use that option but it would seem like changing my use of Facter to $operatingsystem would retain the default behavior if --no-check-prereqs is passed.

Comment (MosesMendoza) [inline install.rb]: Hi @shawnferry - just changing the use of `Facter[...` to $operatingsystem should work. Feels a little less intuitive to me since $operatingsystem hasn't been defined yet but it's not worth the additional effort to substantially revise the options I suppose.

Comment (shawnferry) [inline install.rb]: done, I'll squash the commit and force an update when we are done. Losing the history makes it all harder.

**Comment by shawnferry** [comment]
@MosesMendoza when you are happy with everything I'll squash any new commits

**Comment by MosesMendoza** [comment]
thanks @shawnferry - further review to do on this one just trying to make time for it :)

File: lib/puppet/provider/service/smf.rb [hunk 4]
```diff
@@ -129,6 +129,14 @@ def status
       return
     end
 
+    # Check to see if the service exists
+    cmd = Array[command(:svccfg), "select", @resource[:name]]
+    output = Puppet::Util::Execution.execute(
+        cmd, :combine => true, :failonfail => false)
+    if $CHILD_STATUS.exitstatus != 0
+      info output
+    end
```
Comment (MosesMendoza) [inline lib/puppet/provider/service/smf.rb]: @shawnferry while reviewing this part I realized it appears as if this provider doesn't distinguish between a non-existent service from a stopped/disabled and absent services and can't ensure => absent. I.e., `foo` does not actually exist:
```
-bash-4.1# puppet resource service foo
service { 'foo':
  ensure => 'stopped',
  enable => 'false',
}
```

I filed [PUP-6860](https://tickets.puppetlabs.com/browse/PUP-6860) to address this.

Separate from that, can you pull this existence check logic out into its own method in the provider? I'd like to fix the aforementioned bug and anticipate being able to use it if you could.

Second, I think the way we try to do this now is to reference the command as a method, and use Puppet's logging methods rather than the logging functions (though I admit I'm not 100% sure on that last one). ie:
```
begin
  svccfg(:select, @resource[:name])
rescue Puppet::ExecutionFailure => detail
  Puppet.debug("service was not found, svccfg returned: #{detail.inspect}")
  return false
end
true
```

Comment (shawnferry) [inline lib/puppet/provider/service/smf.rb]: Puppet.debug == debug :: Puppet.info == info; I can fully specify the call if that is the style.

I don't think I can reference the method without the rescue which seems like extra work. as opposed to passing fail_on_fail => false

File: lib/puppet/provider/service/smf.rb [hunk 5]
```diff
@@ -123,6 +123,14 @@ def service_states
...
   end
 
+  def exists?
+    # non-zero exit status is a non-existent service
+    Puppet::Util::Execution.execute(
+      [command(:svccfg), "select", @resource[:name]],
+      :failonfail => false, :combine => true
+    )[1] == 0
+  end
+
```
Comment (shawnferry) [inline lib/puppet/provider/service/smf.rb]: I think this resolves PUP-6860; it could be extracted into
```Ruby
svccfg ...
rescue
```
but I'm not sure it is better.

File: lib/puppet/provider/service/smf.rb [hunk 6]
```diff
@@ -135,8 +143,8 @@ def status
       states = service_states
       state = states[1] == "-" ? states[0] : states[1]
     rescue Puppet::ExecutionFailure
-      info "Could not get status on service #{self.name}"
-      return :stopped
+      debug "Could not get status on service #{self.name} #{$!}"
+      return :absent
```
Comment (shawnferry) [inline lib/puppet/provider/service/smf.rb]: Strictly speaking I'm not sure that we should ever reach this code anymore but I'm not sure. Is there a way to get here without discovering the service doesn't exist via exists? 

If it is reached it also adds the complexity that :absent isn't a valid service state. 
Such that if status is :absent but the requested state is :stopped the end result is a change on every execution that fails to take an absent service to a stopped state.

However, I think that is all addressed via the addition of exists? which then seems to "do the right thing(TM)"

Comment (MosesMendoza) [inline lib/puppet/provider/service/smf.rb]: @shawnferry - my apologies. I led you astray with asking you to define the `exists?` method. I was mistaken in that it turns out the `service` type isn't actually `ensurable` - which means the `exists?` method (as far as I can tell) won't ever be called/used (unless we use it ourselves as a helper for some other task, and we effectively already have that via a failure of the `service_states` method). The only other example of a service provider defining `exists?` is `launchd` and a quick check seems to indicate that isn't used either. I think by simply changing line 147 to `return :absent` we have what we need to solve the problem of puppet printing `ensure => stopped` for non-existent services, and the `exists?` method can be removed.

This means we actually do still reach this rescue block whenever a service doesn't exist. However, I don't think that's a problem. Puppet won't issue a change in that case, but rather a failure to change. Which is what we want.. I.e., with non-existent service `foo`, with your change to `status` method:

```
-bash-4.1# puppet resource service foo
service { 'foo':
  ensure => 'absent',
}
-bash-4.1# puppet resource service foo ensure=stopped
Error: Could not stop Service[foo]: Execution of '/usr/sbin/svcadm disable -s foo' returned 1: svcadm: Pattern 'foo' doesn't match any instances
Error: /Service[foo]/ensure: change from absent to stopped failed: Could not stop Service[foo]: Execution of '/usr/sbin/svcadm disable -s foo' returned 1: svcadm: Pattern 'foo' doesn't match any instances
service { 'foo':
  ensure => 'absent',
}
-bash-4.1# puppet apply -e " service { 'foo' : ensure => stopped } " --detailed-exitcodes
Notice: Compiled catalog for solaris11.localdomain in environment production in 0.33 seconds
Error: Could not stop Service[foo]: Execution of '/usr/sbin/svcadm disable -s foo' returned 1: svcadm: Pattern 'foo' doesn't match any instances
Error: /Stage[main]/Main/Service[foo]/ensure: change from absent to stopped failed: Could not stop Service[foo]: Execution of '/usr/sbin/svcadm disable -s foo' returned 1: svcadm: Pattern 'foo' doesn't match any instances
Notice: Applied catalog in 0.08 seconds
-bash-4.1# echo $?
4
```

Comment (shawnferry) [inline lib/puppet/provider/service/smf.rb]: @MosesMendoza dropped exists. If that's good I'll squash the commits unless you now use the new-ish "squash commits" option on merge.

File: lib/puppet/provider/user/user_role_add.rb [hunk 7]
```diff
@@ -26,8 +27,22 @@
     value !~ /\s/
   end
 
+  def shell=(value)
```
Comment (shawnferry) [inline lib/puppet/provider/user/user_role_add.rb]: Yeah, that is a somewhat questionable fallback and it does look like it should "just work"™ in the case of defining the flag.

The answer is apparently; it originally came from a customer and our notes say it was duplicated but don't give an example.  We've been carrying it since Puppet 3.x

Comment (MosesMendoza) [inline lib/puppet/provider/user/user_role_add.rb]: After sitting on it a bit i like the explicitness of this change. +1 to leaving as-is.

@shawnferry if you want to squash commits to per-ticket I think this is good to go

Comment (shawnferry) [inline lib/puppet/provider/user/user_role_add.rb]: @MosesMendoza squashed

**Comment by MosesMendoza** [comment]
👍

**Comment by MosesMendoza** [comment]
I am running this through an ad-hoc acceptance run and will merge on success

**Comment by MosesMendoza** [comment]
passed an acceptance run against tests impacted by this change
